kythe/docs/rfc/2966.md

```diff
@@ -0,0 +1,331 @@
+standard interface that a tool must implement in order to qualify as a Kythe
+analyzer.
+
+At a high level, an _analyzer_ is a program that consumes Kythe compilation
+records as input, and emits `kythe.proto.Entry` protocol buffer messages as
+output. However, the base definition of Kythe did not specify how this interface
+is to be implemented concretely. That leaves the implementor a great deal of
+flexibility, but makes it unnecessarily difficult to integrate analyzers that
+are implemented in different languages.
+
+The purpose of this proposal is to specify an explicit wire-level protocol for
+communicating compilation records to, and collecting entries from, a Kythe
+analyzer. Some important goals of this proposal are:
+
+1.  To decouple analyzer implementations from the storage representations of
+    compilation records and entry messages. This avoids the need to update
+    analyzers when storage formats change, and makes it more practical to run
+    analyzers in a sandbox.
+
+2.  To make it easier to run analyzers as standalone tools, and reduce or
+    eliminate the need to implement separate wrapper tools for testing vs.
+    production.
+
+3.  To reduce friction for analyzers written in languages that do not have good
+    support for efficient manipulation of binary data, or which require FFI to
+    get good performance.
+
+## Common Terminology
+
+-   A **compilation record** represents a single unit of analysis, corresponding
+    to a single execution of a compiler tool in statically-typed languages. The
+    record consists of a **compilation unit**, stored as a
+    `kythe.proto.CompilationUnit` protocol buffer message, together with the
+    contents of all the **required input** files mentioned by the compilation
+    unit.
+
+-   An **entry** is a datum representing a graph fragment defined by the
+    [Kythe schema](http://kythe.io/docs/schema), as a `kythe.proto.Entry`
+    protocol buffer message.
+
+-   A **Kythe analyzer** is any program that consumes compilation records and
+    emits Kythe entries. More generally, we may use the term **analyzer** to
+    denote a program that consumes compilation records and emits _any_ kind of
+    uniform data; however, this document is only concerned with Kythe analyzers.
+
+-   A **driver** is a program that implements the server side of the interface
+    defined in this document, providing access to the analysis environment for
+    an analyzer program.
+
+## Protocol Description
+
+-   An analyzer communicates with a driver via a pair of reliable simplex byte
+    streams such as Unix pipes, provided by the operating environment. An input
+    stream delivers data from the driver to the analyzer, and an output stream
+    receives output from the analyzer to the driver. In a POSIX environment,
+    these streams correspond to stdin and stdout.
+
+-   The data on each channel are partitioned into consecutive **frames**. Each
```

Comment:
In light of this and your other comment below, I'll rework this section a bit.

Reply:
I've given this some more thought, and am now convinced we should not version the framing discipline at all, and should confine versioning to the embedded protocol.

There are a couple arguments in favour this view:

1. Allowing the framing discipline not only to vary, but also to be negotiated dynamically, adds unnecessary complexity and allows encoding variations (which we are trying to get rid of). Having to support variation at this level adds indelible complexity to the drivers and opens up a choice to analyzer implementors that ultimately does not add much benefit.

2. A length+body framing is already one of the simplest available disciplines. Any variations can only be more complex, or differently-encoded. Given that we've already targeted a text-based protocol, changing the encoding doesn't bring us any clear advantages.

I considered omitting framing entirely, but that's tricky to implement in some languages. The Go JSON decoder can handle a "raw" stream of JSON blobs, but it causes troubles for TypeScript and Dart. It also makes error-detection and recovery harder.  We could use a trailing separator instead of a length prefix, but I think length-prefixed is more friendly—only a small input buffer is required to decode length tags, rather than a flex-buffer and chunk-handling to scan for a delimiter. Nevertheless, we could revisit that if there's some other advantage I missed.